### PR TITLE
[Performance] Reduce allocations in `PENamedTypeSymbol.EnsureTypeParametersAreLoaded`

### PR DESCRIPTION
**Pull request by Youssef1313**
- Fast path for zero arity. Just use empty immutable array directly.
    - This code path shouldn't need to use any of the interlocked APIs since the empty array is always the same instance.
- Use ArrayBuilder instead of an array. Converting an array to immutable array involves allocating another array and copying from the old to the new one. By using ArrayBuilder, a single array is only ever created.
- **We could go further and add two more specialized code paths for one element and two element cases, and use `ImmutableArray.Create(element1)` and `ImmutableArray.Create(element1, element2)` respectively, which would be even more efficient than ArrayBuilder. I'm not sure though if it's worth doing it that way.**

### CODE REVIEW

**Comment by Youssef1313** [comment]
@AlekseyTs

**Review by 333fred — APPROVED** [review]
Special casing doesn't really seem worth it to me for this case. If future perf traces indicate that it would be a real improvement, we can look at it at that point.

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PENamedTypeSymbol.cs [hunk 1]
```diff
@@ -2658,19 +2658,26 @@ private void EnsureTypeParametersAreLoaded()
...
+                        _lazyTypeParameters = ImmutableArray<TypeParameterSymbol>.Empty;
+                        return;
+                    }
+
                     var moduleSymbol = ContainingPEModule;
 
                     // If this is a nested type generic parameters in metadata include generic parameters of the outer types.
                     int firstIndex = _genericParameterHandles.Count - _arity;
 
-                    TypeParameterSymbol[] ownedParams = new TypeParameterSymbol[_arity];
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PENamedTypeSymbol.cs:2666]: @333fred avoiding this unnecessary allocation on every type does seem like a nice to have.

Comment (333fred) [reply src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PENamedTypeSymbol.cs:2666]: Huh?

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PENamedTypeSymbol.cs [hunk 2]
```diff
@@ -2658,19 +2658,26 @@ private void EnsureTypeParametersAreLoaded()
             {
                 if (_lazyTypeParameters.IsDefault)
                 {
+                    if (_arity == 0)
+                    {
+                        _lazyTypeParameters = ImmutableArray<TypeParameterSymbol>.Empty;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PENamedTypeSymbol.cs:2663]: >_lazyTypeParameters = ImmutableArray<TypeParameterSymbol>.Empty; [](http://example.com/codeflow?start=24&length=64)

Strictly speaking, this assignment is not guaranteed to be atomic. Please use ```ImmutableInterlocked.InterlockedInitialize```.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 1)

**Comment by AlekseyTs** [comment]
> This code path shouldn't need to use any of the interlocked APIs since the empty array is always the same instance.

The interlocked API is used to make the assignment atomic, it really doesn't matter whether we are trying to assign the same instance underneath or not. What we are trying to avoid is partial assignment.

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PENamedTypeSymbol.cs (hunk 2 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PENamedTypeSymbol.cs:2663]: As an alternative, we could check arity and assign this way inside the constructor. There is a guarantee that there is no race with any other thread,

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PENamedTypeSymbol.cs:2663]: I would prefer that approach myself.

Comment (Youssef1313) [reply src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PENamedTypeSymbol.cs]: > I would prefer that approach myself.

I assume "that" refers to the second comment? https://github.com/dotnet/roslyn/pull/70766/commits/2456a067124af5b4fce55bf813b11922b4901bb5

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 3)

**Comment by AlekseyTs** [comment]
@Youssef1313 Thank you for the contribution.

**Pull request merged**